pdns/packetcache.hh

```diff
@@ -23,80 +23,137 @@
        + the OPT RR rdlen (2)
-       = 16
+       = 15
     */
-    if(ntohs(dh->arcount)==1 && (pos+16) < packetSize) {
-      char* optionBegin = nullptr;
-      size_t optionLen = 0;
-      /* skip the final empty label (1), the qtype (2), qclass (2) */
-      /* root label (1), type (2), class (2) and ttl (4) */
-      int res = getEDNSOption(const_cast<char*>(reinterpret_cast<const char*>(p)) + 14, end - (p + 14), EDNSOptionCode::ECS, &optionBegin, &optionLen);
-      if (res == 0) {
-        skipBegin = optionBegin;
-        skipEnd = optionBegin + optionLen;
-        if (ecsBegin != nullptr && ecsEnd != nullptr) {
-          *ecsBegin = optionBegin - packet.c_str();
-          *ecsEnd = *ecsBegin + optionLen;
-        }
+    const struct dnsheader* dh = reinterpret_cast<const struct dnsheader*>(packet.data());
+    if (ntohs(dh->qdcount) != 1 || ntohs(dh->ancount) != 0 || ntohs(dh->nscount) != 0 || ntohs(dh->arcount) != 1 || (pos + 15) >= packetSize) {
+      if (packetSize > pos) {
+        currentHash = burtle(reinterpret_cast<const unsigned char*>(&packet.at(pos)), packetSize - pos, currentHash);
       }
+      return currentHash;
     }
-    if (skipBegin > p) {
-      ret = burtle(reinterpret_cast<const unsigned char*>(p), skipBegin-p, ret);
+
+    currentHash = burtle(reinterpret_cast<const unsigned char*>(&packet.at(pos)), 15, currentHash);
+    /* skip the qtype (2), qclass (2) */
+    /* root label (1), type (2), class (2) and ttl (4) */
+    /* already hashed above */
+    pos += 13;
+
+    const uint16_t rdLen = ((static_cast<unsigned char>(packet.at(pos)) * 256) + static_cast<unsigned char>(packet.at(pos + 1)));
+    /* skip the rd length */
+    /* already hashed above */
+    pos += 2;
+
+    if (rdLen > (packetSize - pos)) {
+      if (pos < packetSize) {
+        currentHash = burtle(reinterpret_cast<const unsigned char*>(&packet.at(pos)), packetSize - pos, currentHash);
+      }
+      return currentHash;
     }
-    if (skipEnd < end) {
-      ret = burtle(reinterpret_cast<const unsigned char*>(skipEnd), end-skipEnd, ret);
+
+    uint16_t rdataRead = 0;
+    uint16_t optionCode;
+    uint16_t optionLen;
+
+    while (pos < packetSize && rdataRead < rdLen && getNextEDNSOption(&packet.at(pos), rdLen - rdataRead, optionCode, optionLen)) {
+      if (optionLen > (rdLen - rdataRead - 4)) {
+        if (packetSize > pos) {
+          currentHash = burtle(reinterpret_cast<const unsigned char*>(&packet.at(pos)), packetSize - pos, currentHash);
+        }
+        return currentHash;
+      }
+
+      bool skip = false;
+      if (optionCode == EDNSOptionCode::COOKIE) {
+        skip = true;
+      }
+      else if (optionCode == EDNSOptionCode::ECS) {
+        if (skipECS) {
+          skip = true;
+        }
+      }
+
+      if (!skip) {
+        /* hash the option code, length and content */
+        currentHash = burtle(reinterpret_cast<const unsigned char*>(&packet.at(pos)), 4 + optionLen, currentHash);
```

Comment:
An option having the same opcode but different length does not hash the same.
If options are variable length, this would miss some cache hits. But I guess that's ok, since the equality test also does considers them equal. Actually, do variable length options exist?

Reply:
Yes, it was intended for options with the same opcode but different lengths to compare differently, as the content simply can't be the same. That might change, however, if/when we decide to skip EDNS padding in queries. I guess we would need to special case the EDNS padding option to skip the size of the content.

Reply:
I wonder if it would make sense to skip the length of all options right now. It would make sense for EDNS padding and would also mean that we would fully ignore the content of EDNS Client Subnet options, including the size of the source, which seems like a win when we really do want to ignore ECS.

Reply:
That might actually be an issue for ECS source zero [1], which is a way for the client to ask that no EDNS value be sent to authoritative servers. So we don't want to ignore that one, I guess..

[1]: https://tools.ietf.org/html/rfc7871#section-7.1.2